File: pico-statsd/src/main/scala/org/pico/statsd/Inet.java
package org.pico.statsd;

import com.timgroup.statsd.StatsDClientException;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.UnknownHostException;
import java.util.concurrent.Callable;

public class Inet {
    /**
     * Create dynamic lookup for the given host name and port.
     *
     * @param hostname the host name of the targeted StatsD server
     * @param port     the port of the targeted StatsD server
     * @return a function to perform the lookup
     */
    public static Callable<InetSocketAddress> volatileAddressResolution(final String hostname, final int port) {
        return new Callable<InetSocketAddress>() {
            @Override public InetSocketAddress call() throws UnknownHostException {
                return new InetSocketAddress(InetAddress.getByName(hostname), port);
            }
        };
    }

    /**
     * Lookup the address for the given host name and cache the result.
     *
     * @param hostname the host name of the targeted StatsD server
     * @param port     the port of the targeted StatsD server
     * @return a function that cached the result of the lookup
     * @throws Exception if the lookup fails, i.e. {@link UnknownHostException}
     */
    public static Callable<InetSocketAddress> staticAddressResolution(final String hostname, final int port) throws Exception {
        final InetSocketAddress address = volatileAddressResolution(hostname, port).call();
        return new Callable<InetSocketAddress>() {
            @Override public InetSocketAddress call() {
                return address;
            }
        };
    }

    public static Callable<InetSocketAddress> staticStatsDAddressResolution(final String hostname, final int port) throws StatsDClientException {
        try {
            return staticAddressResolution(hostname, port);
        } catch (final Exception e) {
            throw new StatsDClientException("Failed to lookup StatsD host", e);
        }
    }
}

File: pico-statsd/src/main/scala/org/pico/statsd/Inet.scala
package org.pico.statsd

import java.net.{InetAddress, InetSocketAddress, UnknownHostException}
import java.util.concurrent.Callable

import com.timgroup.statsd.StatsDClientException

object Inet {
  /**
    * Create dynamic lookup for the given host name and port.
    *
    * @param hostname the host name of the targeted StatsD server
    * @param port     the port of the targeted StatsD server
    * @return a function to perform the lookup
    */
  def volatileAddressResolution(hostname: String, port: Int): () => InetSocketAddress = { () =>
    new InetSocketAddress(InetAddress.getByName(hostname), port)
  }

  /**
    * Lookup the address for the given host name and cache the result.
    *
    * @param hostname the host name of the targeted StatsD server
    * @param port     the port of the targeted StatsD server
    * @return a function that cached the result of the lookup
    * @throws Exception if the lookup fails, i.e. { @link UnknownHostException}
    */
  @throws[Exception]
  def staticAddressResolution(hostname: String, port: Int): () => InetSocketAddress = {
    val address: InetSocketAddress = volatileAddressResolution(hostname, port)()
    () => address
  }

  @throws[StatsDClientException]
  def staticStatsDAddressResolution(hostname: String, port: Int): () => InetSocketAddress = {
    try {
      staticAddressResolution(hostname, port)
    } catch {
      case e: Exception => throw new StatsDClientException("Failed to lookup StatsD host", e)
    }
  }
}
